Style Engine: add typography and color to frontend

ramonjd opened the pull request:
## What?

Experimental PR to see if and how we can port backend features added in https://github.com/WordPress/gutenberg/pull/40332 to the frontend in relation to color and typography

## Why?
To unify the frontend and backend. 

It might not work out since styles and classnames aren't added in the same callback 🤷 

## How?

Adding color and typography CSS rules to the existing ruleset (spacing).

~The addition of a new method `getClassnames` to generate block classnames based on preset attribute values passed to the style engine something similar to `var:preset|color|burnedToast`~

## Testing Instructions

Insert a group block into the post editor and tweak the typography and color styles.

Here's some test HTML

<details>

<summary>Block code</summary>

```html
<!-- wp:group {"style":{"color":{"text":"#dd1d1d"},"typography":{"fontSize":"2.08rem","fontStyle":"normal","fontWeight":"500","textTransform":"uppercase","lineHeight":2,"letterSpacing":"23px"},"spacing":{"padding":{"top":"45px","right":"45px","bottom":"45px","left":"45px"},"margin":{"top":"58px","bottom":"58px"}}},"backgroundColor":"secondary"} -->
<div class="wp-block-group has-secondary-background-color has-text-color has-background" style="font-style:normal;color:#dd1d1d;margin-top:58px;margin-bottom:58px;padding-top:45px;padding-right:45px;padding-bottom:45px;padding-left:45px;font-size:2.08rem;font-weight:500;letter-spacing:23px;line-height:2;text-transform:uppercase"><!-- wp:paragraph -->
<p>I love chocolate raisins</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

```

</details>

Check that the block appears as it should in the editor and the frontend.



### Unit tests

```bash
# style engine tests
npm run test-unit packages/style-engine/src/test/index.js 

# global styles output
npm run test-unit components/global-styles/test/use-global-styles-output.js
```
More to come

ramonjd committed:
Returning inline styles by default.

ramonjd committed:
Adding some dev notes for later brain

ramonjd committed:
Adding a couple of typography styles
Updating tests
Refactor styles directory

ramonjd committed:
Adding text decoration
Fixing camel case style prop name

ramonjd committed:
Adding typography properties letterSpacing.ts and textTransform.ts

ramonjd committed:
Adding color.text and typography.fontWeight

ramonjd commented on packages/style-engine/src/styles/typography/fontSize.ts:
```diff
@@ -0,0 +1,32 @@
+
+const fontSize = {
+	name: 'fontSize',
+	generate: ( style: Style, options: StyleOptions ) => {
```
Since most of the new styles share the same generate logic, it might save a few bytes to create a general `getGenerateCallback` utility method.

Premature optimization? You decide!

ramonjd committed:
Refactor generic rule generator

ramonjd committed:
Experiment: adding classname generator

ramonjd commented on packages/style-engine/src/index.ts:
```diff
@@ -62,3 +65,20 @@ export function getCSSRules(
+ *
+ * @return string[] An array of classnames.
+ */
+export function getClassnames( style: Style ): string[] {
```
I think we could probably think about/ship this functionality separately. 

Just adding it here to evaluate the concept.

andrewserong replied:
I like the concept! I think I'd lean toward including it in this PR, it's looking like a good direction to me.

youknowriad replied:
Who's responsible for providing the styles for these classnames, while I think having a function that return "state" classnames is fine, I might think that these classnames shouldn't be necessary to style things properly and if we can have all the logic in `generate` function it's a bit better.

youknowriad replied:
I guess we can't replace classnames with CSS variables yet in `generate` because that would require block deprecations for now. It might be fine to keep these out for now with a caviat that we may change these APIs later (to fully fulfill the style engine's potential :P)

Also, I don't see this function used in the "style" hook, should we replace the custom logic there with this function (like we did with the inline styles)? 

For Global Styles generation though, it seems that we'll have to implement the CSS variable based styles in generate anyway right?

ramonjd replied:
> Who's responsible for providing the styles for these classnames, 

In the post editor, the styles linked to the `has-*` classnames are ultimately pulled from global settings/presents (generated by class-wp-theme-json-gutenberg.php).

An inline JS script hydrates the editor with settings. The styles are passed down [from <Editor />](https://github.com/WordPress/gutenberg/blob/9a195fca8e0a23af1605830b4128edb116a4964a/packages/edit-post/src/editor.js#L160) and [transformed/output in <EditorStyles />](https://github.com/WordPress/gutenberg/blob/efaa86b900f8707a1ae8c11ebf591fde808f24aa/packages/block-editor/src/components/editor-styles/index.js#L71).

Then in [hooks/color.js the classnames](https://github.com/WordPress/gutenberg/blob/6de16981c7fef3db7cc4a9d6cfcb0ad85c916c62/packages/block-editor/src/hooks/color.js#L247) for the block wrapper are compiled and added to the block props.

Is that what you meant by "responsible"?

The last step is what I had in mind for the style engine to take over initially, only because we're doing it in the backend. I was playing around with generating the `has-*` classnames in hooks/color.js via the style engine this morning. It works okay, but I get the feeling I'm building "around" existing logic, so not sure yet.

> while I think having a function that return "state" classnames is fine, I might think that these classnames shouldn't be necessary to style things properly and if we can have all the logic in generate function it's a bit better.

Now I'm writing about it, it might be a better approach to have the style engine generate the CSS styles at the [transformed/output stage mentioned above](https://github.com/WordPress/gutenberg/blob/efaa86b900f8707a1ae8c11ebf591fde808f24aa/packages/block-editor/src/components/editor-styles/index.js#L71) and forget about block classnames for now.

> I guess we can't replace classnames with CSS variables yet in generate because that would require block deprecations for now. 

Yeah, I did have a shot at putting the CSS vars inline... and it worked 🎉  but the console exploded in my face 😄 

> Also, I don't see this function used in the "style" hook, should we replace the custom logic there with this function (like we did with the inline styles)?

I tested it out this morning, and have things half-working. There are some inexplicable fixture-related errors, but I'll push anyway so can eyeball the code.

> For Global Styles generation though, it seems that we'll have to implement the CSS variable based styles in generate anyway right?

Yeah for sure. That will be its own challenge. It makes me think that I should leave the block classname alone for now and rather devote the time into centralizing _style_ generation.

ramonjd replied:
> It makes me think that I should leave the block classname alone for now and rather devote the time into centralizing style generation.

I've pushed my work so far in https://github.com/WordPress/gutenberg/pull/40665/commits/393fac42c75ad97c5d397255fd4f9aefa9cefc25 so we can look.

youknowriad replied:
> Yeah, I did have a shot at putting the CSS vars inline... and it worked 🎉 but the console exploded in my face 😄

But isn't this the right thing to do anyway. It will solve both "block style generation" and "global styles generation" at the same time with the same style engine. 

I agree that we need to find a way to either:

 1- Don't use the style engine yet for colors and such in the hooks/color.js for now and keep the existing logic there
 2- Have some kind of flag per block to say: this block is ready for full style engine style generation which "for me" means:
     - do not serialize inline styles in the saved content
     - rely on the JS style engine for the "edit" function (injected somewhere in hooks/color)
     - rely on the PHP style engine for the frontend.

Maybe it's too soon for 2, I don't know.

---
https://github.com/WordPress/gutenberg/commit/393fac42c75ad97c5d397255fd4f9aefa9cefc25 is fine if we really want to keep getClassnames to centralize the "classname" generation for blocks based on styles, but for me, this is not really the style engine, or more precisely, not the main purpose of the style engine.

I hope that makes a bit of sense :)

ramonjd replied:
> Don't use the style engine yet for colors and such in the hooks/color.js for now and keep the existing logic there

Yeah, what this PR is attempting might be too much too soon. There's probably more value in trying to get style/css generation right across blocks/global styles.

> rely on the PHP style engine for the frontend.

It would be fabulous for the PHP to take care of all frontend styles. Not sure about inline CSS saved in the editor though... We could maybe start over at layout.php for now.

> want to keep getClassnames to centralize the "classname" generation for blocks based on styles, but for me, this is not really the style engine, or more precisely, not the main purpose of the style engine.

I get you. My idea was to take as much decision making about classnames and styles away from the editor as possible. Classname generation might be a bit premature I'd admit that. 

I'd be fine with skipping this functionality in the JS in this PR. Maybe even revisiting the backend functionality too.

 I'm still holding onto the hope that it'll all be clearer after we experiment with things. 🤞 Am I crazy? 😄 

> I hope that makes a bit of sense :)

Yes, it's given us a lot to think about, thanks a lot!

andrewserong replied:
> Maybe it's too soon for 2, I don't know.

Thanks for laying out a bit of how that could work — I think that sounds like something to pursue in a separate PR (perhaps once we've got all the block supports implemented server-side?) but I really like the idea of a per-block flag to opt-in for the style engine to render out the styles server-side on the site frontend. We could also use that flag to conditionally ignore style validation errors as a gentle way of gradually introducing the work from https://github.com/WordPress/gutenberg/pull/37942 (your PR to ignore style validation errors) so we can avoid issues with deprecations.

youknowriad replied:
Thanks for clarifying 👍 I'm confident this is going in the right direction ❤️

ramonjd commented:
Will fix the failing JS tests at some point 😇

andrewserong commented on packages/style-engine/src/index.ts:
```diff
@@ -62,3 +65,20 @@ export function getCSSRules(
+			classNames.push( ...definition.getClassNames( style ) );
+		}
+	} );
+	return uniq( classNames );
```
I know we're already using lodash here (and might not be able to remove it in the shorter-term), but instead of `uniq` I was wondering if we could use a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) for `classNames`, and then return it as an array via `return Array.from( classNames )`?

ramonjd replied:
That would work! Thanks for the tip.

andrewserong commented on packages/style-engine/src/styles/color/background.ts:
```diff
@@ -0,0 +1,44 @@
+	},
+	getClassNames: ( style: Style ) => {
+		const classNames = [];
+		const styleValue: string | undefined = get( style, [
```
Can we get away with optional chaining here instead of using `get`?

ramonjd replied:
Good call. Probably everywhere too. It was a mindless copy paste from the existing code.

andrewserong commented on packages/style-engine/src/styles/color/gradient.ts:
```diff
@@ -0,0 +1,44 @@
+	},
+	getClassNames: ( style: Style ) => {
+		const classNames = [];
+		const styleValue: string | undefined = get( style, [
```
Same as above, can we get away with using optional chaining here?

andrewserong commented on packages/style-engine/src/styles/color/text.ts:
```diff
@@ -0,0 +1,39 @@
+	},
+	getClassNames: ( style: Style ) => {
+		const classNames = [];
+		const styleValue: string | undefined = get( style, [
```
Same as above, can we get away with using optional chaining here?

andrewserong commented on packages/style-engine/src/styles/color/background.ts:
```diff
@@ -0,0 +1,44 @@
+import type { Style, StyleOptions } from '../../types';
+import { generateRule, getSlugFromPreset } from '../utils';
+
+const text = {
```
Tiny typo: looks like this should be `background`.

ramonjd replied:
🤦

andrewserong commented on packages/style-engine/src/styles/color/gradient.ts:
```diff
@@ -0,0 +1,44 @@
+import type { Style, StyleOptions } from '../../types';
+import { generateRule, getSlugFromPreset } from '../utils';
+
+const text = {
```
Tiny typo: looks like this should be `gradient`.

ramonjd replied:
🤦🤦

andrewserong commented on packages/style-engine/src/styles/color/gradient.ts:
```diff
@@ -0,0 +1,44 @@
+		return generateRule(
+			style,
+			[ 'color', 'gradient' ],
+			'gradient',
```
The gradient block support currently stores the value in the `background` CSS property, so I think we'd need to update this line to:

```suggestion
			'background',
```

andrewserong commented on packages/style-engine/src/index.ts:
```diff
@@ -27,20 +27,23 @@ export function generate( style: Style, options: StyleOptions ): string {
 	const groupedRules = groupBy( rules, 'selector' );
 	const selectorRules = Object.keys( groupedRules ).reduce(
 		( acc: string[], subSelector: string ) => {
+			const hasSelector = subSelector !== 'undefined';
```
Is this line (and the one two lines below) missing a `typeof`?

ramonjd replied:
Thanks for raising this. I wasn't sure how to cleanly deal with it.

When there's an undefined object key, `'undefined'` (a string) becomes the key.

```js
const test = { undefined: 2 }
console.log( test['undefined'] ); // 2
```

So... that's why it is how it is 😄

andrewserong replied:
Thanks for the explanation! I must have been reading too fast, and glossed over the `Object.keys()` call, and that the `groupedRules` object is a result of calling `groupBy` — as you mention, since `rules` might have a `selector` attribute of `undefined`, it now makes sense why they're grouped under an `undefined` key 👍

This might be a bit nit-picky given that the logic here works well in testing, but I'm wondering if it'd be worth having a separate condition for handling when we want to return inline styles versus selector-based CSS, mostly because we won't want to mix-and-match (e.g. output containing both an inline style _and_ one wrapped in a selector). The idea would be to try to keep the logic for handling inline styles versus selector-based styles a little more separate. I'm imagining something like the following right at the top of the `generate` function:

```TS
	const rules = getCSSRules( style, options );

	// If no selector is provided, treat generated rules as inline styles to be returned as a single string.
	if ( ! options?.selector ) {
		const inlineRules: string[] = [];
		rules.forEach( ( rule ) => {
			inlineRules.push( `${ kebabCase( rule.key ) }: ${ rule.value };` );
		} );
		return inlineRules.join( ' ' );
	}

	const groupedRules = groupBy( rules, 'selector' );

	// ... the rest of the function
```

What do you think?

ramonjd replied:
> The idea would be to try to keep the logic for handling inline styles versus selector-based styles a little more separate. 

This makes perfect sense to me. It should makes things easier to read and work with in the future. Thanks for thinking about and for the example!

andrewserong commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -220,11 +220,8 @@ function getStylesDeclarations( blockStyles = {} ) {
 
 	// The goal is to move everything to server side generated engine styles
 	// This is temporary as we absorb more and more styles into the engine.
-	const extraRules = getCSSRules( blockStyles, { selector: 'self' } );
```
Just making sure I understand — the idea here is that we treat the absence of specifying a selector as "give me the values that are allowed to be used as inline styles", so we don't need to pass it `'self'`? If so, that makes sense to me 👍

ramonjd replied:
>  the idea here is that we treat the absence of specifying a selector as "give me the values that are allowed to be used as inline styles", so we don't need to pass it 'self'?

Yeah, I though 'self' didn't really communicate much so, in the absence of `options.selector` we'd just return the concatenated style values.

andrewserong commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -251,6 +248,12 @@ export function addSaveProps(
 		}
 	} );
 
+	// @TODO
+	// Experimental notes:
+	// Maybe the style engine could do the classname work of `addSaveProps` in packages/block-editor/src/hooks/color.js here?
+	// And addSaveProps in packages/block-editor/src/hooks/font-family.js
+	// And add it to props.className?
+
```
That makes good sense to me! As you mention further down, though, we could look at that in a separate PR if it winds up causing this one to grow too much.

andrewserong reviewed (commented):
Nice work @ramonjd! This is looking like a really solid direction to me, there's some good parallels with the server-side implementation and I think it's starting to really highlight the value of us having the style engine as a separate package with common patterns for how each of the rules are implemented.

I've left a bunch of comments (mostly just some nits), but it's looking like a good path forward to me. Let me know if there's anything in particular you'd like me to test in further detail!

ramonjd committed:
Fixing bonkers typos because I copy and paste like I just don't care. It's my PR and I'll fail if I want to.
Preferencing `uniq` for new Set() spread
Updating tests.

ramonjd committed:
Enable style engine for backgroundColor

ramonjd committed:
Regenerate pullquote fixture since the style engine changes the order of the inline styles rules. Nothing major.

andrewserong commented on test/integration/fixtures/blocks/core__pullquote__custom-colors.serialized.html:
```diff
@@ -1,3 +1,3 @@
 <!-- wp:pullquote {"textAlign":"right","className":"is-style-default","style":{"color":{"background":"#3486aa","text":"#ffefef"},"border":{"color":"#045679"}}} -->
-<figure class="wp-block-pullquote has-text-align-right is-style-default has-border-color has-text-color has-background" style="background-color:#3486aa;border-color:#045679;color:#ffefef"><blockquote><p>pullquote</p><cite>citation</cite></blockquote></figure>
+<figure class="wp-block-pullquote has-text-align-right is-style-default has-border-color has-text-color has-background" style="border-color:#045679;background-color:#3486aa;color:#ffefef"><blockquote><p>pullquote</p><cite>citation</cite></blockquote></figure>
```
Just adding a brain dump here (there aren't any issues, but I was curious): I was wondering about the potential implications of the order of style values being rendered differently, given that this test fixture needed to be updated.

From manual testing of using the following markup, it looks like the block validation doesn't care about the order of the style values, just that the actual style is equal. Here's the markup I used with this PR applied (the version of this fixture prior to this PR):

```
<!-- wp:pullquote {"textAlign":"right","style":{"color":{"background":"#3486aa","text":"#ffefef"},"border":{"color":"#045679"}},"className":"is-style-default"} -->
<figure class="wp-block-pullquote has-text-align-right is-style-default has-border-color has-text-color has-background" style="background-color:#3486aa;border-color:#045679;color:#ffefef"><blockquote><p>pullquote</p><cite>citation</cite></blockquote></figure>
<!-- /wp:pullquote -->
```

With that markup copy / pasted into the code editor view with this PR applied, I logged out values from [`getStyleProperties`](https://github.com/wordpress/gutenberg/blob/03fd2b68e670cf3a071550db550458ba6c732000/packages/blocks/src/api/validation/index.js#L384) (where block validation splits the style string into an object of key / value pairs) to confirm that the keys of the object are in a different order for the actual vs expected versions of the block:

<img width="521" alt="image" src="https://user-images.githubusercontent.com/14988353/166850356-c87a2dca-ff40-468d-95dd-5234934ed9ee.png">

And then, the [isEqual check](https://github.com/wordpress/gutenberg/blob/03fd2b68e670cf3a071550db550458ba6c732000/packages/blocks/src/api/validation/index.js#L416) is happy about this since it's comparing objects instead of a simple string comparison. Phew! 😁

I mostly just wanted to confirm that we can happily change the order of rendering of style properties and not worry about having to add deprecations or anything like that for existing blocks / block supports (or worry about the order in which the style engine is invoked when generating styles). Good stuff!

ramonjd replied:
Thanks for double-checking @andrewserong 

I also was wondering about this (and pasted the previous order in the code editor too), but fortunately you had already done most of the hard work in https://github.com/WordPress/gutenberg/pull/39039 !

That change saved me a lot of time, so thank you then and now 🙇

andrewserong commented on packages/style-engine/src/index.ts:
```diff
@@ -62,3 +65,21 @@ export function getCSSRules(
+ *
+ * @param  style Style object.
+ *
+ * @return string[] An array of classnames.
```
Tiny typo: I think we don't need to specify the return type here? (I noticed it's repeated in the generated docs)

```suggestion
 * @return An array of classnames.
```

ramonjd replied:
Thanks! I'll build the docs later 😄

ramonjd committed:
Update packages/style-engine/src/index.ts

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd commented:
Just to note that this PR breaks link color style output. The presets aren't parsed. 

So a block like this, 

```html
<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"var:preset|color|vivid-red"}}}}} -->
    <p class="has-link-color"><a href="https://test.com,">Test</a></p>
<!-- /wp:paragraph -->
```

The output is:

```css
.editor-styles-wrapper .wp-elements-3 a{
    color: var:preset|color|vivid-red; /* Should be var(--wp--preset--color--vivid-red)*/
}
```


Looking at this now 👀

ramonjd committed:
Add constants
Updated inline docs
Making the generateRule signature the same as generateBoxRules

ramonjd committed:
Refactoring generate method to split inline and selector logic
Removing styles color text support because we haven't yet come up a with a way to deal with link color styles for elements. We need to reconcile the way we generate text colors with link css var colors and the logic in hooks/style.js
Add constants
Updated inline docs
Making the generateRule signature the same as generateBoxRules

ramonjd committed:
Remove TODO comment. It's something we can do in the future, but it doesn't need to pollute the codebase.

ramonjd commented:
> Just to note that this PR breaks link color style output. The presets aren't parsed.

@andrewserong I've removed support for `styles.color.text` in this PR until we work out how to deal with element styles.

The link text color value needs to be converted from `var:preset|color|slug` to `var(--wp--preset--color--slug)` .

This is fine in itself however the style objects for links and text colors, at least the ones passed to `getInlineStyles()`, are the same.

At [this line ](https://github.com/WordPress/gutenberg/blob/1b8f991d9c0e9400a06af81c76cbaadc113e8078/packages/block-editor/src/hooks/style.js#L121) of `compileElementsStyles()` we pass `{ color: { text: 'value' } }`, which is extracted from `style.elements` further down [here](https://github.com/WordPress/gutenberg/blob/1b8f991d9c0e9400a06af81c76cbaadc113e8078/packages/block-editor/src/hooks/style.js#L341).

The full object would be `{ elements: { link: { color: { text: 'value' } } } }`

Since the link text color and regular text color objects are the same at this point, the style engine doesn't know that it should parse `var:preset|styleContext|slug` values for the link only.

One possibility is to parse **all** style values  for `var:`-like values and return CSS var style rules the time. 

We can do that and it will work (I tested it).

However if we do that, we might need to rethink the way we generate classnames, given that the current "rule" for background colors and other styles is use `var:preset|color|slug` to generate and return a classname, e.g., `has-$slug-color`, only and not return `CSS var style rules`.

Long story short, the style engine would need to know when to return a CSS var and when not to for a particular style value.

Given it's the style hook that's giving the style object a haircut, I think a neater way would be to pass the entire style object, including the `elements` object, to the style engine so we can do it explicitly for each style.

This means we might need to concentrate and/or refactor hooks/style.js, which would entail a bit of faffing around in another PR.

What do you reckon? Sorry if that doesn't make sense. I've been thinking about it for too long.

Can you think of a way of dealing with `var:` values that need to be converted into CSS vars?

ramonjd committed:
Reinstating color and now parsing style value for `var:`, whereby we'd return CSS vars all the time.

ramonjd commented:
> One possibility is to parse all style values for var:-like values and return CSS var style rules the time.

> We can do that and it will work (I tested it).

I've done that in [2cf8698](https://github.com/WordPress/gutenberg/pull/40665/commits/2cf86987490013420cfa3774c42ac9f398482d30)

The consequence is that we'll parse all style values for `var:...` strings and return a CSS var.

Maybe it's a better way to start off and we can deal with classname + element stuff later, the backend notwithstanding 🤷

ramonjd committed:
A bit of defensive coding to check for a string type

andrewserong commented:
> However if we do that, we might need to rethink the way we generate classnames, given that the current "rule" for background colors and other styles is use var:preset|color|slug to generate and return a classname, e.g., has-$slug-color, only and not return CSS var style rules.

For the other colors that might reference a variable (e.g. `background` and `gradient`), they're currently handled outside of the style object, so that appears to be how they skip outputting a CSS var style rule, unless I'm missing something?

> One possibility is to parse all style values for var:-like values and return CSS var style rules the time.

That seems like the most straightforward way to handle it to me, since it looks like that's the current behaviour in `getInlineStyles` where each value is run through [compileStyleValue](https://github.com/wordpress/gutenberg/blob/1b8f991d9c0e9400a06af81c76cbaadc113e8078/packages/block-editor/src/hooks/style.js#L97)?

Apologies, I very well might be missing quite a bit of context here!

ramonjd commented:
> That seems like the most straightforward way to handle it to me, since it looks like that's the current behaviour in getInlineStyles where each value is run through [compileStyleValue](https://github.com/wordpress/gutenberg/blob/1b8f991d9c0e9400a06af81c76cbaadc113e8078/packages/block-editor/src/hooks/style.js#L97)?

Thanks for trying to decipher my gibberish @andrewserong 

I ended up doing this anyway. 

What I was trying to express with all the Github comment anguish was this: in the backend `var:preset|*` values are used to build classnames only.

So 

```php
    'color' => array(
       'background' => 'var:preset|color|burned-toast' 
   )
```

Tells the style engine to return `has-burned-toast-background-color` and not `background-color: var(--wp--preset--color--burned-toast)`.

See https://github.com/WordPress/gutenberg/pull/40332#issuecomment-1098818214 for some context.

This logic doesn't exist in the frontend right now. Maybe it doesn't have to, and all my hand-wringing is for nothing.

Besides, we don't really have to deal with it yet since we're not using the [getClassNames](https://github.com/WordPress/gutenberg/pull/40665/files#diff-99efe2b4db43717aa11901f762109d4dbde79a5d9c8968d9ff61e91d7c39452aR83) method.

My guess is that it will come out in the wash as we progress.

andrewserong commented:
> What I was trying to express with all the Github comment anguish was this: in the backend `var:preset|*` values are used to build classnames only.

Ah, I see now! Thanks for the explanation, [this is the line](https://github.com/wordpress/gutenberg/blob/ef1f44994ecf68f4466ff4d8b5ab8c2f768ed567/packages/style-engine/class-wp-style-engine.php#L194) where you're skipping generating the CSS rule if it starts with `var` 👍

> My guess is that it will come out in the wash as we progress.

Yes, it's a little tricky to work out which way we should go here. I _think_ I'd lean toward seeing how far we can get with the current approach, even if it creates a slight gap between frontend and backend, and then revisit when we explicitly go to incorporate the [Elements block support](https://github.com/wordpress/gutenberg/blob/d36ee9642b5d64e08cf088839dd3179d8d310429/lib/block-supports/elements.php#L113) — that'll wind up involving dealing with non-inline styles, too, so it becomes a bit of a rabbithole to figure out.

One of the other differences we still have between front and backend is that in the frontend code we're also calling the style engine from the global styles output, which we haven't started doing in the backend yet. So probably good to see if we can handball some of these questions to follow-up PRs where we can dig in, in isolation?

ramonjd commented:
> So probably good to see if we can handball some of these questions to follow-up PRs where we can dig in, in isolation?

Agree. Once again, thanks for helping out with the thinking here. 🙇

ramonjd committed:
Testing out generating block color classnames using the style engine.

ramonjd committed:
Also passing the style color value to `getClassnames` so it will add the generic class name if required.
Regenerating fixtures due to the re-ordering of color classnames.

ramonjd committed:
Reverting `getClassnames`. It's a bigger change and should be looked at in another PR.

ramonjd committed:
Fix typo in fixture

andrewserong commented on packages/blocks/src/api/constants.js:
```diff
@@ -127,10 +130,12 @@ export const __EXPERIMENTAL_STYLE_PROPERTY = {
 	fontWeight: {
```
Directly above this line, is `fontStyle` missing an opt-in to `useEngine`?

ramonjd replied:
Good call! 👍

andrewserong commented on packages/style-engine/src/styles/utils.ts:
```diff
@@ -46,3 +90,47 @@ export function generateBoxRules(
+ *
+ * @return generated styles.
+ */
+export function getSlugFromPreset(
```
Is this function being used anywhere yet? I couldn't see it being called from anywhere, so wondered if it was part of the `getClassnames` behaviour that was removed. Probably fine to leave it in, just curious!

ramonjd replied:
Another good call! 😄 

This was an artefact from my tests with a specific callback for elements and also returning classnames. 

Since that we're not doing any of that yet, I removed it with the intention of looking at it in another PR, after the [backend](https://github.com/WordPress/gutenberg/pull/40987) is done.

andrewserong commented on packages/style-engine/src/styles/typography/index.ts:
```diff
@@ -0,0 +1,111 @@
+import type { Style, StyleOptions } from '../../types';
+import { generateRule } from '../utils';
+
+const fontFamily = {
```
Apologies, if this is a naive question — I haven't looked it up yet 😅 — do we need the font family support to output inline styles like this, or is it always implemented by classname? (Just wondering if this support is needed here)

ramonjd replied:
You're right to question this. I wasn't sure either.

I added it based on the backend implementation, which does check for a custom font-family in the block attribute style object.

Here's the block supports typography.php [line after the style engine migration](https://github.com/WordPress/gutenberg/blob/ef1f44994ecf68f4466ff4d8b5ab8c2f768ed567/lib/block-supports/typography.php#L110) and [before the migration](https://github.com/WordPress/gutenberg/blob/6de16981c7fef3db7cc4a9d6cfcb0ad85c916c62/lib/block-supports/typography.php#L108).

So I thought to myself, _"maybe there's a case for a custom font family of which I'm not aware"._ Then I stopped thinking. 😄 

But the [editor/JS block supports hook explicitly checks only for the attributes.fontFamily](https://github.com/WordPress/gutenberg/blob/6de16981c7fef3db7cc4a9d6cfcb0ad85c916c62/packages/block-editor/src/hooks/font-family.js#L118) value, which seems to assume that it's always a preset/classname thing.

I'll take it out to keep it inline with the current JS implementation.

Thank you!

andrewserong reviewed (commented):
Thanks for tidying this up @ramonjd, I think this is looking very close to merge to me.

Tested with the following markup (and making lots of tweaks in the editor):

<details>

<summary>Block markup</summary>

```html
<!-- wp:group {"style":{"color":{"gradient":"linear-gradient(135deg,rgb(252,185,0) 18%,rgb(255,105,0) 100%)"}}} -->
<div class="wp-block-group has-background" style="background:linear-gradient(135deg,rgb(252,185,0) 18%,rgb(255,105,0) 100%)"><!-- wp:heading {"fontFamily":"system-font"} -->
<h2 class="has-system-font-font-family">Heading</h2>
<!-- /wp:heading -->

<!-- wp:paragraph {"style":{"typography":{"fontSize":"0.53rem"}}} -->
<p style="font-size:0.53rem">A paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"typography":{"fontStyle":"italic","fontWeight":"800","textTransform":"capitalize","lineHeight":"5","letterSpacing":"13px"}},"fontSize":"medium"} -->
<p class="has-medium-font-size" style="font-style:italic;font-weight:800;letter-spacing:13px;line-height:5;text-transform:capitalize">This is a paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#f66d6d"}}},"color":{"background":"#6d0e0e","text":"#f0f0f0"}},"fontSize":"large"} -->
<p class="has-text-color has-background has-link-color has-large-font-size" style="color:#f0f0f0;background-color:#6d0e0e">Another paragraph with <a href="http://www.imdb.com">a link</a>.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:read-more {"content":"Read more","style":{"typography":{"textDecoration":"underline"}}} /-->
``` 
</details>

<img width="555" alt="image" src="https://user-images.githubusercontent.com/14988353/168762143-e3b0ce1f-399c-4046-86f1-c819c8fed347.png">

✅ Color is output correctly
✅ Background color is output correctly
✅ Font size is output correctly
❓ Font style is output correctly, but it looks like we need to add the `useEngine` flag to it (tests well manually if I change that flag)
✅ Font weight is output correctly
✅ Line height is output correctly
✅ Text decoration is output correctly
✅ Text transform is output correctly
✅ Letter spacing is output correctly
❓ Font family doesn't appear to be opted-in — was that intentional? I assume so, as it's currently implemented via class names in the editor for e.g. Heading block? I left a comment to discuss further, I ran out of time today to look it up myself 😅

Happy to do more testing / a final review tomorrow. Thanks again! 🙇

ramonjd committed:
Flag fontStyle with useEngine: true so that it uses the style engine.
Remove the handler for fontFamily to reflect the current situation in the editor whereby no custom fontFamily style values are expected.
Removed unused function `getSlugFromPreset`
Added tests for elements.link.color.text preset style values.

andrewserong reviewed (approved):
Thanks for making the updates @ramonjd! I've just re-tested and that's looking good now.

✅ Font style is output correctly
✅ Font family is no longer included
✅ Included utils functions are used, and no strays left behind, and logic in `getCSSVarFromStyleValue` matches that used in `computeStyleValue` in `hooks/style.js` 👍

This LGTM now! 🎉
